akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala

```diff
@@ -658,26 +662,29 @@ object Http extends ExtensionId[HttpExt] with ExtensionIdProvider {
   final case class IncomingConnection(
     localAddress: InetSocketAddress,
     remoteAddress: InetSocketAddress,
-    flow: Flow[HttpResponse, HttpRequest, Unit]) {
+    flow: Flow[HttpResponse, HttpRequest, Promise[Option[Option[InetSocketAddress]]]]) {
 
     /**
      * Handles the connection with the given flow, which is materialized exactly once
      * and the respective materialization result returned.
      */
-    def handleWith[Mat](handler: Flow[HttpRequest, HttpResponse, Mat])(implicit fm: Materializer): Mat =
-      flow.joinMat(handler)(Keep.right).run()
+    def handleWith[Mat](handler: Flow[HttpRequest, HttpResponse, Mat])(implicit fm: Materializer): Mat = {
+      val fillMeInAndUserMat = flow.joinMat(handler)(Keep.both).run()
+      fillMeInAndUserMat._1.success(Some(Some(remoteAddress)))
+      fillMeInAndUserMat._2
+    }
```

Comment:
>On the other hand, you could also map the materialized value Promise at the same point where you would set the attribute.

Indeed. I wonder what the impact of that would be.

>Still, as @ktoso said, using the materialized value makes the API more error prone for all consumers who don't want to set the RemoteAddress which is an optional feature.

Absolutely. But using the Attribute means you first of all have to know that it exists and that you may want to set it. Right now it becomes a hidden setting. Do you agree with that assessment?